R/filter-nm-data.R

```diff
@@ -0,0 +1,193 @@
+#' Helper for reading in and parsing the `$DATA` record.
+#'
+#' @note This cannot be used for _modifying_ a `$DATA` record.
+#' @noRd
+#' @keywords internal
+read_data_record <- function(.mod){
+  data_recs <- get_records(.mod, "data")
+  n_data <- length(data_recs)
+  if(n_data !=1){
+    recs_fmt <- purrr::map_chr(data_recs, function(rec) rec$format())
+    rlang::abort(
+      c(
+        glue::glue("Expected a single data record, but found {n_data}:\n\n"),
+        recs_fmt
+      )
+    )
+  }
+  data_recs[[1]]$parse()
+  return(data_recs[[1]])
+}
+
+
+#' Extract IGNORE or ACCEPT options from a data record
+#' @param data_rec an `nmrec_record_data` object
+#' @inheritParams translate_nm_expr
+#' @noRd
+get_data_filter_exprs <- function(data_rec, type = c("ignore", "accept")){
+  type <- match.arg(type)
+
+  # Extract & format IGNORE/ACCEPT options
+  opts <- purrr::keep(data_rec$values, function(val){
+    inherits(val, "nmrec_option_value") && identical(val[["name"]], type)
+  })
+
+  exprs <- purrr::map_chr(opts, function(val){
+    gsub("\\(|\\)", "", unquote_filename(val$value))
+  })
+
+  return(exprs)
+}
+
+#' Function to translate NONMEM operators to R operators
+#' @noRd
+#' @keywords internal
+translate_nm_operator <- function(expr) {
+  expr <- gsub("\\.EQ\\.|\\.EQN\\.", "==", expr)
+  expr <- gsub("\\.NE\\.|\\.NEN\\.|/=", "!=", expr)
+  expr <- gsub("\\.LT\\.", "<", expr)
+  expr <- gsub("\\.LE\\.", "<=", expr)
+  expr <- gsub("\\.GT\\.", ">", expr)
+  expr <- gsub("\\.GE\\.", ">=", expr)
+
+  # Handle single `=` only when it's not part of `==`, `!=`, `<=`, `>=`
+  expr <- gsub("(?<=[^=!<>])=(?=[^=!<>])", "==", expr, perl = TRUE)
+  return(expr)
+}
+
+#' Function to invert ignore expressions
+#' @noRd
+#' @keywords internal
+invert_operator <- function(expr) {
+  expr <- dplyr::case_when(
+    grepl("==", expr) ~ gsub("==", "!=", expr),
```

Comment:
I added it to the location you mentioned for the `gsub` expression. It broke things when I attempted to add it to `translate_nm_operator`, but let me know if you were thinking of including this in the `grepl` expression instead

Reply:
Ok to just leave it. But look at the `pattern` there .. it's just looking for a literal string. So you can use `fixed = TRUE` in that case can be faster and safer. 

In `translate_nm_operator()`, there's all sorts of back ticks in there, but it's not needed in the last 4. You do need them in the first two, but you could render all this using `fixed = TRUE` and it would run the same (even faster). 

```r
translate_nm_operator <- function(expr) {
  # EQN and NEN are available after NONMEM 7.3
  # expr <- gsub("\\.EQ\\.|\\.EQN\\.", "==", expr) # regular expression used here, but you don't have to 
  # expr <- gsub("\\.NE\\.|\\.NEN\\.|/=", "!=", expr) # regular expression used here, but not required
  expr <- gsub("\\.LT\\.", "<", expr)
  expr <- gsub("\\.LE\\.", "<=", expr)
  expr <- gsub("\\.GT\\.", ">", expr)
  expr <- gsub("\\.GE\\.", ">=", expr)
```